Errors in test suite(pending) and `getUnixTime`(resolved in 0.4.6) results for windows

Hi, testing and trying to fix the package build for win 7, we found that:
* The test suite fails in windows ( windows 7 and 10, cabal-2.4.1.0, ghc-8.6.5):
```console
Test suite spec: RUNNING...

UnixTime
  formatUnixTime
    behaves like the model FAILED [1]
  parseUnixTimeGMT & formatUnixTimeGMT
    inverses the result
      +++ OK, passed 100 tests.
    inverses the result (2) FAILED [2]
  addUnixDiffTime & diffUnixTime
    invrses the result
      +++ OK, passed 100 tests.
  UnixTime Storable instance
    peek . poke = id
      +++ OK, passed 100 tests.

Failures:

  test\\UnixTimeSpec.hs:44:13:
  1) UnixTime.formatUnixTime behaves like the model
       Falsifiable (after 4 tests):
         UnixTime {utSeconds = 1635071885, utMicroSeconds = 542638}
       expected: "Sun, 24 Oct 2021 12:38:05 +0200"
        but got: "Sun, 24 Oct 2021 12:38:05 "

  To rerun use: --match "/UnixTime/formatUnixTime/behaves like the model/"

  test\\UnixTimeSpec.hs:53:9:
  2) UnixTime, parseUnixTimeGMT & formatUnixTimeGMT, inverses the result (2)
       Falsified (after 4 tests):
         UnixTime {utSeconds = 1635071885, utMicroSeconds = 542638}

  To rerun use: --match "/UnixTime/parseUnixTimeGMT & formatUnixTimeGMT/inverses the result (2)/"
```
* Otoh The value of `utSeconds` returned  by `getUnixTime` seems to be incorrect. The program:
```haskell
module Main where

import Data.UnixTime
import qualified GHC.IO.Encoding
import qualified System.IO

main = do
  GHC.IO.Encoding.setLocaleEncoding System.IO.utf8
  putStrLn "Testing unix time"
  t <- getUnixTime
  let t' = UnixTime {  utSeconds = 4294967295, utMicroSeconds =999999 }
  putStrLn . show $ t
  m <- formatUnixTime mailDateFormat t
  putStrLn . show $ m
  putStrLn . show $ t'
  m' <- formatUnixTime mailDateFormat t'
  putStrLn . show $ m'
```
has this output
```console
Testing unix time
UnixTime {utSeconds = 207173602136727, utMicroSeconds = 48236}
"?, -1 ? 1899 -1:-1:-1 "
UnixTime {utSeconds = 4294967295, utMicroSeconds = 999999}
"Sun, 07 Feb 2106 07:28:15 +0100"

```

I cant test it with stack, it throws an error that sometimes is throwed by cabal too:
```
PS D:\dev\ws\haskell\unix-time> stack test
unix-time-0.4.5: build (lib + test)
Preprocessing library for unix-time-0.4.5..
Building library for unix-time-0.4.5..
Preprocessing test suite 'spec' for unix-time-0.4.5..
Building test suite 'spec' for unix-time-0.4.5..
Linking .stack-work\dist\e626a42b\build\spec\spec.exe ...
D:\dev\ws\haskell\unix-time\.stack-work\dist\e626a42b\build/libHSunix-time-0.4.5-IBIazFuQIJnJhbGD18ZUwr.a(strptime.o):strptime.c:(.text+0x1064): referencia a `timegm' sin definir
collect2.exe: error: ld returned 1 exit status
`gcc.exe' failed in phase `Linker'. (Exit code: 1)
....
```
Using stack-1.9.3, default project stack.yaml, windows 10
When cabal throws the error is due a bad config.h generation in the configure step: the default `config.h.in` definitions are not replaced by the values computed in configure.

--- a/Data/UnixTime/Conv.hs
+++ b/Data/UnixTime/Conv.hs
@@ -71,6 +71,7 @@ parseUnixTimeGMT fmt str = unsafePerformIO $
 -- This is a wrapper for strftime_l().
 -- 'utMicroSeconds' is ignored.
 -- The result depends on the TZ environment variable.
+--
 
 formatUnixTime :: Format -> UnixTime -> IO ByteString
 formatUnixTime fmt t =

--- a/cbits/conv.c
+++ b/cbits/conv.c
@@ -50,7 +50,7 @@ char *set_tz_utc() {
     char *tz;
     tz = getenv("TZ");
 #if defined(_WIN32)
-    _patch_setenv("TZ", "", 1);
+    _patch_setenv("TZ", "UTC", 1);
 #else
     setenv("TZ", "", 1);
 #endif

--- a/cbits/strftime.c
+++ b/cbits/strftime.c
@@ -522,7 +522,8 @@ label:
 #ifdef ALTZONE
 					diff = -altzone;
 #else /* !defined ALTZONE */
-					continue;
+					// Fix the daylight saving time, see #54.
+					diff = -(_timezone - 3600 * t->tm_isdst);
 #endif /* !defined ALTZONE */
 #endif /* !defined TM_GMTOFF */
 				if (diff < 0) {

--- a/cbits/win_patch.c
+++ b/cbits/win_patch.c
@@ -109,9 +109,9 @@ int _patch_setenv(const char *var, const char *val, int _ovr) {
     int len = varlen + vallen + 2;
     char *sname = (char *)malloc(len);
     strcpy(sname, var);
-    sname[strlen(var)] = '=';
+    sname[varlen] = '=';
     strcpy(sname + varlen + 1, val);
-    sname[len + 1] = '\0';
+    sname[varlen + vallen + 1] = '\0';
     int r = _putenv(sname);
     free(sname);
     return r;

--- a/cbits/win_patch.c
+++ b/cbits/win_patch.c
@@ -100,18 +100,26 @@ const struct lc_time_T   _C_time_locale = {
 };
 
 
-int _patch_setenv(const char *var, const char *val, int ovr) {
-    BOOL b = SetEnvironmentVariableA(var, val);
-    if (b) {
-        return 0;
-    } else {
-        return 1;
+int _patch_setenv(const char *var, const char *val, int _ovr) {
+    if (val == NULL) {
+        return _patch_unsetenv(var);
     }
+    int varlen = strlen(var);
+    int vallen = strlen(val);
+    int len = varlen + vallen + 2;
+    char *sname = (char *)malloc(len);
+    strcpy(sname, var);
+    sname[strlen(var)] = '=';
+    strcpy(sname + varlen + 1, val);
+    sname[len + 1] = '\0';
+    int r = _putenv(sname);
+    free(sname);
+    return r;
 }
 
 int _patch_unsetenv(const char *name) {
-    int len = strlen(name);
-    char *sname = (char *)malloc(len + 2);
+    int len = strlen(name) + 2;
+    char *sname = (char *)malloc(len);
     strcpy(sname, name);
     sname[len] = '=';
     sname[len + 1] = '\0';

